### DISCUSSION TITLE
Lesson 6 error about connecting to local host

### DISCUSSION THREAD

**Discussion by SNERUOL** [Cyfrin/foundry-full-course-cu#1240] [Q&A]
Hi!

While following the video I am running into this error message, I was expecting to fill in a private key from Metamask but instead I got this. 
I also searched in the repository, asked Google, asked ChatGPT but still didn't understand what would be the next step.
Does anyone know what to do now? Please help me with this!

![Schermafbeelding 2023-12-24 225834](https://github.com/Cyfrin/foundry-full-course-f23/assets/153216108/88b6fdd9-392a-45ab-9265-ab6423599f61)
![Schermafbeelding 2023-12-24 225816](https://github.com/Cyfrin/foundry-full-course-f23/assets/153216108/0d590470-61ad-4f8b-ac5b-1819d36d66c5)

**Comment by ghost** [comment]
The [`create`](https://book.getfoundry.sh/reference/forge/forge-create) command from Forge expects the path of the contract to deploy, not just its name. Also, the port of your localhost isn't `8545`?

Launch Anvil, copy one of the generated private keys and run the following command:

```bash
forge create src/SimpleStorage.sol:SimpleStorage --rpc-url http://127.0.0.1:7545 --interactive
```
When asked for the pk, paste it. It should automatically deploy the contract.

**Reply by SNERUOL** [reply]
Hi pacelliv,

Thanks a lot for your reply!

1. So to use the create command I have to do what? I don't understand.

2. And I wanted to use Ganache, just like in the video. Not Anvil. What would it be then?
Didn't I put this line that you gave here in my code? It looks the same also ending on --interactive.
Or do I have to write forge create **SRC** <--- so including the src and then on?

Looking forward to your answer.

Regards,
Lourens

**Comment by xyz899** [comment]
### Hi, 
#### Answering your questions here

in order to use the `create` command, you have to provide an RPC URL in order for you to test on it directly in the Command Line  with the path and name of the Contract and an account or private key for you to use for example :

`forge create src/YourContractFilename.sol:contractName --rpc-url http://127.0.0.1:8545 --private-key <YourPrivateKey or ENV variable containing the private key> (or --interactive to copy paste directly your key after validating the command without revealing it)` 

If you use an .env file with for example :

##### .env 
`PRIVATE_KEY = 0x.....5b`

You'll then proceed by making the environnement variable `source .env` and then use `--private-key $PRIVATE_KEY`

and If you want to use Ganache you will need to create a new workspace with accounts, logs, mined block per sec, and get the rpc url and paste it with this argument : `--rpc-url <rpc-url-link>`

![Screenshot 2023-12-25 102011](https://github.com/Cyfrin/foundry-full-course-f23/assets/112932882/319ddaef-bcf5-4748-a150-0ef16216ae9f)

As you can see on this image my rpc url is : `http://127.0.0.1:7545` copy paste and It should be good to go

**Reply by SNERUOL** [reply]
Hi Winter9998,

Thanks a lot for your answer and merry x-mas!

I still do not know how to go on with this. I thought I did put in the values of the rpc url but nog after your message and filling in a path it's still not good. 
Please see the screenshots and I hope you have another answer.

Regards,
Lourens
![Schermafbeelding 2023-12-25 160342](https://github.com/Cyfrin/foundry-full-course-f23/assets/153216108/c1e6d8ff-a377-4884-b587-fdc976ee5dbc)
![Schermafbeelding 2023-12-25 160404](https://github.com/Cyfrin/foundry-full-course-f23/assets/153216108/b26fd1d1-de86-4a93-868f-8f549cfca30b)
![Schermafbeelding 2023-12-25 160353](https://github.com/Cyfrin/foundry-full-course-f23/assets/153216108/65a7415c-cae5-439a-b1d8-43359d3095c4)

**Reply by xyz899** [reply]
### Merry Christmas, wish you best to you and loved ones

#### Primary advise is to read what foundry outputs you at the error, they did a fantastic job and have almost every error covered whether in Foundry's Docs or directly into the Command Line,this is the correct way to write with forge create :

`forge create src/YourContractFilename.sol:contractName`

`forge create` : for the command to create a contract,
`src/YourContractFilename.sol` : the path of the Contract,
`YourContractFilename.sol:contractName` the contractName which is what you forgot to include be sure to add just after the path the contract path with :contractName in your case :

`forge create src/SimpleStorage.sol:SimpleStorage --rpc-url http://127.0.0.1:8545 --private-key <privatekey>` ✅ 

you specified the path with the name of a folder you're already in foundry-simple-storage-f23 there isn't any folder named like this inside your folder in which resides src, lib script, test

**Reply by xyz899** [reply]
Check my answers above, Any typo inside the command line could make your test fail so be aware and good luck out there

**Reply by SNERUOL** [reply]
Hi Winter thanks again for your message!

Yes I read the error text message but I do not understand what it means.
And nog I still don't know how this works. Everything was fine until now to be honest...
I put in the line you wrote, this one:

forge create src/SimpleStorage.sol:SimpleStorage --rpc-url http://127.0.0.1:8545 --private-key <privatekey> ✅
because this was supposed to be the right one, right?
I changed the 8545 (where did you get this from?) to 7545 because that is what I see at the top of the Ganache window as my RPC server number. And at the end I put in the private key from the top line key icon in Ganache.
Still I am getting this error and I searched for it but I couldn't get any wiser...

So to me I am doing what you tell me but ofcourse somewhere I am not doing it right.

![Schermafbeelding 2023-12-25 174209](https://github.com/Cyfrin/foundry-full-course-f23/assets/153216108/1905f1b7-9a28-4446-b64e-9486d151f595)

**Reply by SNERUOL** [reply]
Maybe it will be a good thing to check out the freeCodeCamp BASH tutorial...

**Reply by xyz899** [reply]
No, I think you need computer science in general this is supposedly the basics of command line, urls and so on. Mine is 8545 because I used anvil ( which is directly integrated into the forge binary cli )  for the endpoint and not ganache. Never copy/paste something without knowing how or why.

**Reply by xyz899** [reply]
and if you're not familiar with bash at all or all these tools I would heavily recommend you to start off by doing basic and general courses for computer science, you can't dive in web3 if you aren't aware of web2 and how it works, like webapps, system tools and so on.

**Reply by SNERUOL** [reply]
Yes I understand, thank you!
But could you tell me please how I can  continue with this part?

And before I "copied and pasted" you line of code I checked it to see how it was formed and then already didn't understand why there was a different number. But I thought, Ok just let's try this. So, it wasn't mindless simply copy paste.

**Reply by xyz899** [reply]
> Yes I understand, thank you!
> 
> But could you tell me please how I can  continue with this part?
> 
> 
> 
> And before I "copied and pasted" you line of code I checked it to see how it was formed and then already didn't understand why there was a different number. But I thought, Ok just let's try this. So, it wasn't mindless simply copy paste.

Like I said before sure to learn computer science basic before diving into web3 

Here's your line : 

`forget create src/SimpleStorage.sol:SimpleStorage --rpc-url http://127.0.0.1:7545 --private-key <yourPVKey>`

If you still don't get it, read my previous answers and use the course videos and if you still don't get it stop doing web3 learn web2 and computer science General Terms. then learn web3

**Reply by SNERUOL** [reply]
Thanks a lot for your help! Putting in this line with the private key leads to another new syntax error: 
bash: syntax error near unexpected token `newline'
Ofcourse it is good to learn these errors, but I have not seen them in the video which I think is a bit strange.
I will take a look into the stuff you mentioned, thanks for the advice.

Enjoy the rest of your day!

Regards,
Lourens
